third_party/nvidia/hopper/lib/Transforms/WarpSpecialization/TMEMAlloc1D.cpp

```diff
@@ -0,0 +1,229 @@
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Pass/PassManager.h"
+#include "nvidia/hopper/include/Transforms/Passes.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonNvidiaGPU/IR/Dialect.h"
+
+#define DEBUG_TYPE "nvgpu-1D-tmem-alloc"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+namespace tt = mlir::triton;
+namespace ttg = mlir::triton::gpu;
+namespace ttng = ::mlir::triton::nvidia_gpu;
+namespace mlir {
+
+// Wrapper class to hold the context for handling
+// 1D TMEM Allocation.
+class TMEM1DAllocator {
+private:
+  OpBuilder &builder;
+  Block *globalStart;
+  int numWarps;
+  // Intermediate info to minimize code reuse across functions.
+  tt::ExpandDimsOp expandedInput = nullptr;
+  ttng::TMEMAllocOp allocOp = nullptr;
+
+public:
+  TMEM1DAllocator(OpBuilder &builder, Block *globalStart)
+      : builder(builder), globalStart(globalStart),
+        numWarps(ttg::lookupNumWarps(&globalStart->front())) {}
+
+private:
+  void copyAttrs(Operation *oldOp, Operation *newOp) {
+    // Right now we just copy over ttg.partition per
+    // the example.
+    // TODO: Should we copy over loop information?
+    auto partitionAttr = oldOp->getAttr("ttg.partition");
```

Comment:
The task only specified `ttg.partition` so I copied that example. I'm not sure how we want to approach `loop.stage` and `loop.cluster`. Once I've done a deep dive on loop scheduling here I might update this.